compiler/hash-source/src/constant.rs

```diff
@@ -0,0 +1,291 @@
+//! be passed around each stage of the compiler and can just
+//! be interned, and accessed when needed.
+
+use std::fmt::Display;
+
+use dashmap::DashMap;
+use fnv::FnvBuildHasher;
+use hash_utils::counter;
+use lazy_static::lazy_static;
+use num_bigint::BigInt;
+
+use crate::identifier::{Identifier, CORE_IDENTIFIERS};
+
+/// Interned float constant which stores the value of the float, and
+/// an optional `type ascription` which is a suffix on the literal
+/// describing which float kind it is, either being `f32` or `f64`.
+#[derive(Debug, Clone, Copy)]
+pub struct FloatConstant {
+    /// Raw value of the float
+    pub value: f64,
+    /// If the constant contains a type ascription, as specified
+    /// when the constant is declared, e.g. `32.4f64`
+    pub suffix: Option<Identifier>,
+}
+
+counter! {
+    name: InternedFloat,
+    counter_name: INTERNED_FLOAT_COUNTER,
+    visibility: pub,
+    method_visibility:,
+}
+
+impl Display for FloatConstant {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.value)?;
+
+        if let Some(suffix) = self.suffix {
+            write!(f, "{suffix}")?;
+        }
+
+        Ok(())
+    }
+}
+
+impl Display for InternedFloat {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", CONSTANT_MAP.lookup_float_constant(*self))
+    }
+}
+
+/// Value of the [IntConstant], this kind be either the `inlined`
+/// variant where we just fallback to using `u128` for small sized
+/// integer constants, and then in the unlikely scenario of needing
+/// more than a [u128] to represent the constant, we will then
+/// fallback to [BigInt].
+#[derive(Debug, Eq, PartialEq, Hash)]
+pub enum IntConstantValue {
+    /// For small values, we store inline
+    Small(u128),
+    /// For bigger values, we just store a pointer to the `BigInt`
+    Big(Box<BigInt>),
+}
+
+impl From<BigInt> for IntConstantValue {
+    fn from(value: BigInt) -> Self {
+        let bits = value.bits();
+
+        // We want to see if we can fit this big-int in a `u128` so we can just copy it
+        // directly
+        if bits <= 128 {
+            let data = (<BigInt as TryInto<i128>>::try_into(value).unwrap()).to_be_bytes();
+            IntConstantValue::Small(u128::from_be_bytes(data))
+        } else {
+            IntConstantValue::Big(Box::new(value))
+        }
+    }
+}
+
+/// Interned literal constant which stores the raw value of the
+/// integer and an optional `type ascription` which binds the
+/// defined literal value to some ascribed type.
+#[derive(Debug, PartialEq, Eq, Hash)]
+pub struct IntConstant {
+    /// Raw value of the integer. This is stored as either a `u128` which can be
+    /// directly stored as the value which happens in `99%` of cases, if the
+    /// constant is not big enough to store this integer, then we resort to
+    /// using [IntConstant
+    pub value: IntConstantValue,
```

Comment:
This probably shoudn't be pub, someone might use the small variant and treat it as a number rather than bytes..